scipy/optimize/tests/test_zeros.py

```diff
@@ -205,6 +206,307 @@ def test_lru_cached_individual(self, method):
+            zeros._chandrupatla(lambda x: x, [-2, -3], [3, 4, 5])
+        with pytest.raises(ValueError, match=message):
+            zeros._chandrupatla(lambda x: [x[0], x[1], x[1]], [-3, -3], [5, 5])
+
+        message = 'Tolerances must be non-negative scalars.'
+        with pytest.raises(ValueError, match=message):
+            zeros._chandrupatla(lambda x: x, -4, 4, xatol=-1)
+        with pytest.raises(ValueError, match=message):
+            zeros._chandrupatla(lambda x: x, -4, 4, xrtol=None)
+        with pytest.raises(ValueError, match=message):
+            zeros._chandrupatla(lambda x: x, -4, 4, fatol='ekki')
+        with pytest.raises(ValueError, match=message):
+            zeros._chandrupatla(lambda x: x, -4, 4, frtol=None)
+
+        message = '`maxiter` must be a non-negative integer.'
+        with pytest.raises(ValueError, match=message):
+            zeros._chandrupatla(lambda x: x, -4, 4, maxiter=1.5)
+        with pytest.raises(ValueError, match=message):
+            zeros._chandrupatla(lambda x: x, -4, 4, maxiter=-1)
+
+        message = '`callback` must be callable.'
+        with pytest.raises(ValueError, match=message):
+            zeros._chandrupatla(lambda x: x, -4, 4, callback='shrubbery')
+
+    def test_special_cases(self):
+        # Test edge cases and other special cases
+
+        # Test that integers are not passed to `f`
+        # (otherwise this would overflow)
+        def f(x):
+            # assert np.issubdtype(x.dtype, np.floating)
+            return x ** 99 - 1
+
+        res = zeros._chandrupatla(f, -7, 5)
+        assert res.success
+        assert_allclose(res.x, 1)
+
+        # Test that if both ends of bracket equal root, algorithm reports
+        # convergence
+        def f(x):
+            return x**2 - 1
+
+        res = zeros._chandrupatla(f, 1, 1)
+        # assert res.success
+        assert_equal(res.x, 1)
+
+        def f(x):
+            return 1/x
+
+        with np.errstate(invalid='ignore'):
+            res = zeros._chandrupatla(f, np.inf, np.inf)
+        assert res.success
+        assert_equal(res.x, np.inf)
+
+        # Test maxiter = 0. Should do nothing to bracket.
+        def f(x):
+            return x**3 - 1
+
+        bracket = (-3, 5)
+        res = zeros._chandrupatla(f, *bracket, maxiter=0)
+        assert res.xl, res.xr == bracket
+        assert res.nit == 0
+        assert res.nfev == 2
+        assert res.status == -2
+        assert res.x == -3  # best so far
+
+        # Test maxiter = 1
+        res = zeros._chandrupatla(f, *bracket, maxiter=1)
+        assert res.success
+        assert res.status == 0
+        assert res.nit == 1
+        assert res.nfev == 3
+        assert_allclose(res.x, 1)
+
+        # Test scalar `args` (not in tuple)
+        def f(x, c):
+            return c*x - 1
+
+        res = zeros._chandrupatla(f, -1, 1, args=3)
+        assert_allclose(res.x, 1/3)
+
+        # # TODO: Test zero tolerance
```

Comment:
Not in this PR. None of the `root_scalar` methods allow zero tolerance. I want to fix this, but I think I can provide a general solution for all bracketing rootfinders in `root_scalar`: make the bracket very small (e.g. one or a handful of floating point spacings), evaluate the function at all the values, and the argmin is the root.